File: Programing/HigherOrderFunction/Exercise2/AnswerJava/Calculator.java
import java.util.function.*;
import java.util.HashMap;

class Calculator {
    private HashMap<Operator, IntBinaryOperator> operators
	= new HashMap<Operator, IntBinaryOperator>();
    
    public void addOperator(Operator operatorType, IntBinaryOperator operator) {
	operators.put(operatorType, operator);
    }

    public int calculate(Operator operatorType, int operand1, int operand2) {
	if (operators.containsKey(operatorType)) {
	    return operators
		.get(operatorType)
		.applyAsInt(operand1, operand2);
	}
	else {
	    throw new IllegalStateException();
	}
    }
}

File: Programing/HigherOrderFunction/Exercise2/AnswerCSharp/Calculator.cs
using System;
using System.Collections.Generic;

namespace Exercise2
{
    class Calculator
    {
        private readonly Dictionary<Operator, Func<int, int, int>> operators
	    = new Dictionary<Operator, Func<int, int, int>>();

        public void AddOperator(Operator operatorType, Func<int, int, int> function)
	{
	    operators[operatorType] = function;
	}

        public int Calculate(Operator operatorType, int operand1, int operand2)
	{
	    if (operators.ContainsKey(operatorType))
	    {
	        return operators[operatorType](operand1, operand2);
	    }
	    else
	    {
	        throw new InvalidOperationException();
	    }
	}
    }
}
